### PR TITLE
add flowbits parse

### PR DESCRIPTION
**Pull request by chuanjiesun**

### CODE REVIEW

File: rule.go [hunk 1]
```diff
@@ -73,6 +75,12 @@ type Metadata struct {
 	Value string
 }
 
+// Flowbits description.
+type Flowbits struct {
```
Comment (duanehoward) [inline rule.go]: The type here should be "Flowbit" as this struct represents a singular flowbit, not multiple

File: rule.go [hunk 2]
```diff
@@ -73,6 +75,12 @@ type Metadata struct {
 	Value string
 }
 
+// Flowbits description.
```
Comment (duanehoward) [inline rule.go]: Please complete this description

File: rule.go [hunk 3]
```diff
@@ -58,6 +58,8 @@ type Rule struct {
 	Vars map[string]*Var
 	// Metas is a slice of Metadata.
 	Metas Metadatas
+	//flowbits
+	Flowbs []*Flowbits
```
Comment (duanehoward) [inline rule.go]: s/Flowbs/Flowbits
Flowbits []*Flowbit

File: rule.go [hunk 4]
```diff
@@ -58,6 +58,8 @@ type Rule struct {
 	Vars map[string]*Var
 	// Metas is a slice of Metadata.
 	Metas Metadatas
+	//flowbits
```
Comment (duanehoward) [inline rule.go]: // Flowbits is a slice of Flowbit

File: rule.go (hunk 1 above)
Comment (duanehoward) [inline rule.go]: additions of any new type should also include a Stringer for writing these back out, and be included in the overall Stringer for the Rule struct as well

File: parser.go [hunk 5]
```diff
@@ -358,7 +358,7 @@ func (r *Rule) option(key item, l *lexer) error {
 		}
 		v.NumBytes = n
 
-		offset, err := strconv.Atoi(parts[1])
+		offset, err := strconv.Atoi(strings.TrimSpace(parts[1]))
```
Comment (duanehoward) [inline parser.go]: This seems to be unrelated to the change to add flowbits? Probably put this in its own PR

Comment (chuanjiesun) [reply parser.go]: > Do you have a specific need for flowbits to be their own struct? If not it may suffice simply to add "flowbits" to the keywords allowed for tags here:
> https://github.com/google/gonids/blob/master/parser.go#L178

flowbits maybe occur more than once,so  deal with it addtionally is better. also additional processing of byte_test keyword might be better,the rule like below contains four byte_test:
**alert tcp any any -> $HOME_NET !$HTTP_PORTS (msg:"ET CURRENT_EVENTS Malformed HeartBeat Request"; flow:established,to_server; content:"|18 03|"; depth:2; byte_test:1,<,4,2; content:"|01|"; offset:5; depth:1; byte_extract:2,3,record_len; byte_test:2,>,2,3; byte_test:2,>,record_len,6; threshold:type limit,track by_src,count 1,seconds 120; flowbits:set,ET.MalformedTLSHB; reference:cve,2014-0160; reference:url,blog.inliniac.net/2014/04/08/detecting-openssl-heartbleed-with-suricata/; reference:url,heartbleed.com/; reference:url,blog.fox-it.com/2014/04/08/openssl-heartbleed-bug-live-blog/; classtype:bad-unknown; sid:2018372; rev:2; metadata:created_at 2014_04_08, updated_at 2014_04_08;)**

**alert ip $HOME_NET any -> [184.106.133.130,184.73.167.34] any (msg:"CNC Shadowserver Reported CnC Server IP group 10"; reference:url,doc.emergingthreats.net/bin/view/Main/BotCC; reference:url,www.shadowserver.org; threshold: type limit, track by_src, seconds 3600, count 1; flowbits:set,ET.Evil; flowbits:set,ET.BotccIP; classtype:trojan-activity; sid:2404009; rev:5417; metadata:affected_product Any, attack_target Any, deployment Perimeter, tag Shadowserver, signature_severity Major, created_at 2012_05_04, updated_at 2019_07_03;)**

Comment (chuanjiesun) [reply parser.go]: > This seems to be unrelated to the change to add flowbits? Probably put this in its own PR

sorry for that, I forgot delete this test code, I will delete it.

File: parser.go [hunk 6]
```diff
@@ -381,6 +381,19 @@ func (r *Rule) option(key item, l *lexer) error {
 		r.Vars[name] = v
 		lastContent := r.Contents[len(r.Contents)-1]
 		lastContent.Options = append(lastContent.Options, &ContentOption{Name: key.value, Value: strings.Join(parts, ",")})
+	case key.value == "flowbits":
+		nextItem := l.nextItem()
+		parts := strings.Split(nextItem.value, ",")
+		tmp_fb := &Flowbit{}
```
Comment (duanehoward) [inline parser.go]: this isn't 'tmp' because this is the pointer that you're appending, just call it 'fb' or similar

File: parser_test.go [hunk 7]
```diff
@@ -832,6 +832,45 @@ func TestParseRule(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "flowbits",
+			rule: `alert http $HOME_NET any -> $EXTERNAL_NET any (msg:"ACTIVEX ATTACK Citrix Access Gateway Plug-in ActiveX Code Execution - SET"; flow:to_server,established; content:"epaq"; http_uri; flowbits:set,ETPRO.Citrix_AGAX; flowbits:noalert; classtype:attempted-user; sid:2803611; rev:4;)`,
```
Comment (duanehoward) [inline parser_test.go]: This appears to be an emerging-threats (ET PRO even based on the flowbit) rule based on the SID. I'd like to keep third party rules out of this test file. Please make this a generic rule that exercises the necessary components of the code without relying on third party (especially paid) detections.

File: rule.go (hunk 4 above)
Comment (duanehoward) [reply rule.go]: This comment still needs to be updated.

File: parser_test.go [hunk 8]
```diff
@@ -832,6 +832,45 @@ func TestParseRule(t *testing.T) {
...
+					
+				},
+				Flowbits: []*Flowbit{
+					&Flowbit{
+						Action: "set",
+						Value: "ETPRO.Citrix_AGAX",
+					},
+					&Flowbit{
+						Action: "",
+						Value: "noalert",
```
Comment (duanehoward) [inline parser_test.go]: The 'action' is noalert. The parser should handle this differently I believe.

Comment (chuanjiesun) [reply parser_test.go]: that's right , action noalert is better

File: parser.go [hunk 9]
```diff
@@ -381,6 +381,19 @@ func (r *Rule) option(key item, l *lexer) error {
 		r.Vars[name] = v
 		lastContent := r.Contents[len(r.Contents)-1]
 		lastContent.Options = append(lastContent.Options, &ContentOption{Name: key.value, Value: strings.Join(parts, ",")})
+	case key.value == "flowbits":
+		nextItem := l.nextItem()
+		parts := strings.Split(nextItem.value, ",")
+		fb := &Flowbit{}
+		if len(parts) == 2{
```
Comment (duanehoward) [inline parser.go]: This code should be refactored to use a guard clauses and return any errors in the flowbits definition, something like this:

https://play.golang.org/p/cSj5NFGLUj4

File: parser_test.go [hunk 10]
```diff
@@ -832,6 +832,45 @@ func TestParseRule(t *testing.T) {
...
+					Ports: []string{"any"},
+				},
+				Destination: Network{
+					Nets:  []string{"$EXTERNAL_NET"},
+					Ports: []string{"any"},
+				},
+				SID:         2803611,
+				Revision:    4,
+				Description: "Flowbits test",
+				Tags: map[string]string{"flow": "to_server,established", "classtype": "attempted-user"},
```
Comment (duanehoward) [inline parser_test.go]: please add newlines between the `flow` and `classtype` entries for easier readability. Alternately remove these, since they're not necessary for the testing of flowbits parsing.

File: rule.go (hunk 4 above)
Comment (duanehoward) [inline rule.go]: This comment still needs updating.

File: parser_test.go [hunk 11]
```diff
@@ -832,6 +832,45 @@ func TestParseRule(t *testing.T) {
...
+						Action: "set",
+						Value: "Citrix_AGAX",
+					},
+					&Flowbit{
+						Action: "noalert",
+						Value: "",
+					},
+				},
+			},
+		},
```
Comment (duanehoward) [inline parser_test.go]: can you add an error test case down near line 901 to test invalid flowbits returning error?

something like alert tcp $HOME_NET any -> $EXTERNAL_NET any (msg:"bad flowbit"; flowbits:invalid_action; ...)

File: parser.go [hunk 12]
```diff
@@ -37,6 +37,16 @@ var escapeRE = regexp.MustCompile(`([()+.'\\])`)
 // metaSplitRE matches string in metadata
 var metaSplitRE = regexp.MustCompile(`,\s*`)
 
+//inSlice will test if a string in a slice
+func inSlice(str string, strings []string) bool {
```
[the remark this answers is not in the document]
Comment (chuanjiesun) [reply parser.go]: I have found that, the code has been deleted.

File: rule.go [hunk 13]
```diff
@@ -58,6 +58,8 @@ type Rule struct {
 	Vars map[string]*Var
 	// Metas is a slice of Metadata.
 	Metas Metadatas
+	// Flowbits is a slice of Flowbit.
```
Comment (duanehoward) [inline rule.go]: please pull from head and handle merge issues.
And please update with Stringers for the new type.

Comment (chuanjiesun) [inline rule.go]: Add parse support for keyword flowbits in rules.

**Comment by chuanjiesun** [comment]
> This all looks good. Please provide the Stringers for the Flowbits type, and add it to the Rule Stringer here https://github.com/google/gonids/blob/master/rule.go#L416

sorry, Where to add stringers type "provide the Stringers for the Flowbits type"?

**Comment by duanehoward** [comment]
> > This all looks good. Please provide the Stringers for the Flowbits type, and add it to the Rule Stringer here https://github.com/google/gonids/blob/master/rule.go#L416
> 
> sorry, Where to add stringers type "provide the Stringers for the Flowbits type"?

Probably around here[0]. You'll see the other String() functions for the other types (like PCRE, Metadata, etc.)

Then you'll need to integrate that String() call as needed into the one for Rule so that when you call `r.String()` you get the correctly formatted rule with all of the portions of the Rule struct.

Does this make sense?

[0] https://github.com/google/gonids/blob/master/rule.go#L414

**Comment by chuanjiesun** [comment]
> > > This all looks good. Please provide the Stringers for the Flowbits type, and add it to the Rule Stringer here https://github.com/google/gonids/blob/master/rule.go#L416
> > 
> > 
> > sorry, Where to add stringers type "provide the Stringers for the Flowbits type"?
> 
> Probably around here[0]. You'll see the other String() functions for the other types (like PCRE, Metadata, etc.)
> 
> Then you'll need to integrate that String() call as needed into the one for Rule so that when you call `r.String()` you get the correctly formatted rule with all of the portions of the Rule struct.
> 
> Does this make sense?
> 
> [0] https://github.com/google/gonids/blob/master/rule.go#L414

thanks, I have add the stringers, please check that.

File: rule.go [hunk 14]
```diff
@@ -419,6 +419,21 @@ func (p PCRE) String() string {
 	return s.String()
 }
 
+//string returns a string for all of the flowbits
```
Comment (duanehoward) [inline rule.go]: All comments need a space following `//` shoudl start with a capital letter and end with a `.` See Effective Go[0] section on Commentary.

// String returns a string for a flowbit.

[0] https://golang.org/doc/effective_go.html#commentary

File: rule.go [hunk 15]
```diff
@@ -419,6 +419,21 @@ func (p PCRE) String() string {
...
+func (fb Flowbit) String() string {
+	var s strings.Builder
+	if inSlice(fb.Action, []string{"noalert", "isset", "isnotset", "set", "unset", "toggle"}) {
```
Comment (duanehoward) [inline rule.go]: In order to reduce indentation you should use guard clauses[0], something like this:

```
if !inSlice(fb.Action, []string{"noalert", "isset", ...}) {
  return ""
}
var s strings.Builder
s.WriteString(fmt.Sprintf("flowbits:%s", fb.Action))
if fb.Value != "" {
  s.WriteString(fmt.Sprintf(",%s", fb.Value))
}
s.WriteString(";")
return s.String()
```  
**

[0] https://medium.com/@scadge/if-statements-design-guard-clauses-might-be-all-you-need-67219a1a981a
** note that I didn't test this code. Please verify.

File: rule.go [hunk 16]
```diff
@@ -412,6 +419,21 @@ func (p PCRE) String() string {
 	return s.String()
 }
 
+//string returns a string for all of the flowbits
+func (fb Flowbit) String() string {
```
Comment (duanehoward) [inline rule.go]: You'll need to add tests for this (see other String() tests)[0]

[0] https://github.com/google/gonids/blob/master/rule_test.go#L502

File: rule.go [hunk 17]
```diff
@@ -448,6 +470,10 @@ func (r Rule) String() string {
 	for k, v := range r.Tags {
 		s.WriteString(fmt.Sprintf("%s:%s; ", k, v))
 	}
+	
+	for _, fb := range r.Flowbits {
```
Comment (duanehoward) [inline rule.go]: Please add a test case that exercises the new flowbits string functionality[0]

[0] https://github.com/google/gonids/blob/master/rule_test.go#L540

**Comment by duanehoward** [comment]
Merged! Thanks!
